src/pynxtools/dataconverter/readers/multi/reader.py

```diff
@@ -0,0 +1,446 @@
+        dims: Optional[Callable[[str, str], List[str]]] = None,
+        entry_name: str = "entry",
+    ):
+        self.special_key_map = {
+            "@attrs": attrs_callback if attrs_callback is not None else self.identity,
+            "@link": link_callback if link_callback is not None else self.link_callback,
+            "@data": data_callback if data_callback is not None else self.identity,
+            "@eln": eln_callback if eln_callback is not None else self.identity,
+        }
+
+        self.dims = dims if dims is not None else lambda *_, **__: []
+        self.entry_name = entry_name
+
+    def link_callback(self, key: str, value: str) -> Dict[str, Any]:
+        """
+        Modify links to dictionaries with the correct entry name.
+        """
+        return {"link": value.replace("/entry/", f"/{self.entry_name}/")}
+
+    def identity(self, _: str, value: str) -> str:
+        """
+        Returns the input value unchanged.
+
+        This method serves as an identity function in case no callback is set.
+        """
+        return value
+
+    def apply_special_key(self, precursor, key, value):
+        """
+        Apply the special key to the value.
+        """
+        return self.special_key_map.get(precursor, self.identity)(key, value)
+
+
+def resolve_special_keys(
+    new_entry_dict: Dict[str, Any],
+    key: str,
+    value: Any,
+    optional_groups_to_remove: List[str],
+    callbacks: ParseJsonCallbacks,
+    suppress_warning: bool = False,
+) -> None:
+    """
+    Resolves the special keys (denoted by "@") through the callbacks.
+
+    Also takes care of the "!" notation, i.e., removes optional groups
+    if a required sub-element cannot be filled.
+    """
+
+    def try_convert(value: str) -> Union[str, float, int, bool]:
+        """
+        Try to convert the value to float, int or bool.
+        If not convertible returns the str itself.
+        """
+
+        if is_integer(value):
+            return int(value)
+        if is_number(value):
+            return float(value)
+        if is_boolean(value):
+            return to_bool(value)
+
+        return value
+
+    def parse_config_value(value: str) -> Tuple[str, Any]:
+        """
+        Separates the prefixes (denoted by "@") from the rest
+        of the value.
+
+        Parameters
+        ----------
+        value : str
+            Config dict value.
+
+        Returns
+        -------
+        Tuple[str, Any]
+            Tuple like (prefix, path).
+
+        """
+        # Regex pattern to match @prefix:some_string
+        pattern = r"!?(@\w+)(?::(.*))?"
+        prefixes = re.findall(pattern, value)
+        if not prefixes:
+            return ("", value)
+        return prefixes[0]
+
+    # Handle non-keyword values
+    if not isinstance(value, str) or "@" not in str(value):
+        if isinstance(value, str):
+            # Handle "!" notation for required fields in optional groups
+            value = value.lstrip("!")
```

Comment:
I think we will never end up in this code part with just a single literal value. Even if we did I think it would be fine to remove the `!` and do no extra check for literal values (would prevent to write literals beginning with `!` though)

Reply:
Ok, now while having a closer look to this it makes sense to remove this check.